opentelemetry_collector/receiver/nginxreceiver/nginx_stats_collector.go

```diff
@@ -0,0 +1,224 @@
+package nginxreceiver
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumerdata"
+	"go.opentelemetry.io/collector/consumer/pdatautil"
+	"go.uber.org/zap"
+
+	"github.com/googlecloudplatform/appengine-sidecars-docker/opentelemetry_collector/receiver/metricgenerator"
+)
+
+// NginxStatsCollector is a struct that generates metrics by polling the nginx status page at statsUrl.
+type NginxStatsCollector struct {
+	consumer consumer.MetricsConsumer
+
+	now       func() time.Time
+	startTime time.Time
+	done      chan struct{}
+	logger    *zap.Logger
+	getStatus func(string) (resp *http.Response, err error)
+
+	exportInterval time.Duration
+	statsUrl       string
+}
+
+// LatencyStats is a struct to parse the latency stats json into.
+type LatencyStats struct {
+	RequestCount int64   `json:"request_count"`
+	LatencySum   int64   `json:"latency_sum"`
+	SumSquares   int64   `json:"sum_squares"`
+	Distribution []int64 `json:"distribution"`
+}
+
+// NginxStats is a struct to parse the nginx stats json into.
+type NginxStats struct {
+	RequestLatency      LatencyStats `json:"request_latency"`
+	UpstreamLatency     LatencyStats `json:"upstream_latency"`
+	WebsocketLatency    LatencyStats `json:"websocket_latency"`
+	LatencyBucketBounds []float64    `json:"latency_bucket_bounds"`
+}
+
+const (
+	defaultExportInterval = time.Minute
+)
+
+// NewNginxStatsCollector creates a new NginxStatsCollector that generates metrics
+// based on nginx stats found by polling the url
+func NewNginxStatsCollector(exportInterval time.Duration, url string, logger *zap.Logger, consumer consumer.MetricsConsumer) *NginxStatsCollector {
+	if exportInterval <= 0 {
+		exportInterval = defaultExportInterval
+	}
+
+	collector := &NginxStatsCollector{
+		consumer:       consumer,
+		now:            time.Now,
+		done:           make(chan struct{}),
+		logger:         logger,
+		exportInterval: exportInterval,
+		statsUrl:       url,
+		getStatus:      http.Get,
+	}
+
+	return collector
+}
+
+// StartCollection starts a go routine that periodically polls nginx for stats and exports metrics based on them.
+func (collector *NginxStatsCollector) StartCollection() {
+	collector.startTime = collector.now()
+
+	go func() {
+		ticker := time.NewTicker(collector.exportInterval)
```

Comment:
nit: missing `ticker.Stop()` invocation -- either as a `defer` statement or after for loop.

Reply:
done